### PR TITLE
Add support for RebootMethod_HALT for Reboot API

### PR DESCRIPTION
**Pull request by vvolam**
#### Why I did it
Add support for HALT reboot method for Reboot gNOI API

#### How I did it
Implement changes for HaltSystem() function to be called when HALT reboot method is invoked.

#### How to verify it
Unit tests

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: test/test_gnoi.py [hunk 1]
```diff
@@ -21,6 +21,17 @@ def test_gnoi_reboot(self):
         assert ret == 0, 'Fail to read counter'
         assert new_cnt == old_cnt+1, 'DBUS API is not invoked'
 
+    def test_gnoi_reboot_halt(self):
+        ret, old_cnt = gnmi_dump('DBUS config reload')
```
Comment (hdwhdw) [inline test/test_gnoi.py]: Do you mind explaining what these dumps (DBUS config reload, DBUS API not invoked") are for? They do not seem related to the test case.

File: gnmi_server/gnoi.go [hunk 2]
```diff
@@ -158,18 +172,29 @@ func (srv *SystemServer) Reboot(ctx context.Context, req *gnoi_system_pb.RebootR
...
-	config_db_json, err := io.ReadFile(fileName)
-	if errors.Is(err, os.ErrNotExist) {
-		fileName = ""
-	}
-	err = RebootSystem(string(config_db_json))
-	if err != nil {
-		return nil, err
+
+	// Check the reboot type
+	if req.GetMethod() == gnoi_system_pb.RebootMethod_HALT {
```
Comment (hdwhdw) [inline gnmi_server/gnoi.go:177]: Maybe using switch statement here to make it easier to add more method?

File: sonic_service_client/dbus_client.go [hunk 3]
```diff
@@ -203,10 +203,9 @@ func (c *DbusClient) HaltSystem() error {
     busPath := c.busPathPrefix + modName
     intName := c.intNamePrefix + modName + ".execute_reboot"
 
-    // Command to execute the system halt
-    command := "sudo reboot"
+    method := 3
```
Comment (hdwhdw) [inline sonic_service_client/dbus_client.go:206]: Can we have a documentation here for what this "3" means?

File: test/test_gnoi.py (hunk 1 above)
Comment (hdwhdw) [inline test/test_gnoi.py]: honest question: what are these gnmi_dump('DBUS config reload') for? They don't seem immediately related to the test.

File: test/test_gnoi.py [hunk 4]
```diff
@@ -22,7 +22,7 @@ def test_gnoi_reboot(self):
         assert new_cnt == old_cnt+1, 'DBUS API is not invoked'
 
     def test_gnoi_reboot_halt(self):
-        ret, old_cnt = gnmi_dump('DBUS config reload')
```
Comment (hdwhdw) [inline test/test_gnoi.py]: I see test case comes from the last commit. Maybe consider squashing this one into the last one?

File: gnmi_server/gnoi.go (hunk 2 above)
Comment (vvolam) [reply gnmi_server/gnoi.go]: Done. Thanks!

File: test/test_gnoi.py (hunk 1 above)
Comment (vvolam) [reply test/test_gnoi.py]: This is a mistake, fixed in latest version!

File: test/test_gnoi.py (hunk 4 above)
Comment (vvolam) [reply test/test_gnoi.py]: I don't have write permission to the repo for squash merge unfortunately.

**Review by hdwhdw — APPROVED** [review]

**Comment by prgeor** [comment]
@vvolam can you add all the code PRs to the PR description?

**Comment by qiluo-msft** [comment]
> 

Do not change a file with only empty lines in diff.

---
Refers to: gnoi_client/gnoi_client.go:353 in 5f81358. [](commit_id = 5f8135848c00ed60c81dc5ac51e06a884312b666, deletion_comment = False)

File: sonic_service_client/dbus_client.go (hunk 3 above)
Comment (qiluo-msft) [reply sonic_service_client/dbus_client.go:207]: Agree, request change here.

**Comment by vvolam** [comment]
> > 
> 
> Do not change a file with only empty lines in diff.
> 
> Refers to: gnoi_client/gnoi_client.go:353 in [5f81358](https://github.com/sonic-net/sonic-gnmi/commit/5f8135848c00ed60c81dc5ac51e06a884312b666). [](commit_id = [5f81358](https://github.com/sonic-net/sonic-gnmi/commit/5f8135848c00ed60c81dc5ac51e06a884312b666), deletion_comment = False)

@qiluo-msft Removed the empty lines change. Thank you.

**Review by hdwhdw — APPROVED** [review]

**Review by ganglyu — DISMISSED** [review]

**Review by qiluo-msft — DISMISSED** [review]

**Comment by vvolam** [comment]
Fixed the merging conflicts. @hdwhdw @ganglyu @qiluo-msft please re-approve.

**Review by hdwhdw — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]
